### ISSUE TITLE
Inviting non-registered users to join a group

### ISSUE THREAD

**Issue by meyerch** [geosolutions-it/unesco-ihp#91]
Hi, 
I noticed an error when inviting non registered users to a group [both the public and dev instance].
The non-registered person receives a link to join the group. 
1st problem: with the said link, we either end up on this page 
![invitation_group1.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46e19a8c2a9682bd86e6b17ca34fff813917653f6b14b1247a80e03cfd183954560e1aa8445c0912adfc8b2ef0b63c738a580becd46b5d66da2f53fde6d640e258f075c95d8fa4bd8246c1ed3c27e68f8a4f3c1d02118bed10.PNG)

or an error page  (see http://ihp-wins-dev.geo-solutions.it/accounts/login/?next=/groups/group/SDG652/invite/ef310eb4eff0f1b110190cbef9dad0b76f0e718b/ for example) 
![invitation_group3.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46e19a8c2a9682bd86e6b17ca34fff813915653f6b14b1247a80e03cfd183954560e1aa8445c0912adfc8b2ef0b63c738a580becd46b5d66da2f53fde6d640e258f075c95d8fa4be8b46c1ed3c27e68f8a4e3e16011f8aed10.PNG)

If the person ends up on the first page and click "Join group", she is not redirected to the registering page. Instead, the persons arrives on this error page:
![invitation_group2.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46e19a8c2a9682bd86e6b17ca34fff813914653f6b14b1247a80e03cfd183954560e1aa8445c0912adfc8b2ef0b63c738a580becd46b5d66da2f53fde6d640e258f075c95d8fa4bd8246c1ed3c27e68f8a4d3b1f051188ed10.PNG)

Can you solve the issue so that people who are invited to join a group and are nto yet registered on the platform can click "join the group" and have the registration page. Once their account is approved by the administrator, those people are automatically member of the group they have agreed to join.

**Comment by ricardogsilva** [comment]
After doing some initial research I've some considerations. 

Joining of a group is handled differently depending on if the invited user is already registered in the site or not:

- If the user is already registered, a group's manager will simply add the user to the group without any need for input from the user. In this case there is no invitation, the user is simply added to the group.

- If the user is not yet registered, the group's manager must invite the user to join the group (and also join the site implicitly). The user is first notified via e-mail that she has been invited to join. Then she must first register with the site - this is a two-step process, since it requires admin approval; Finally, the user must accept the invitation to join the group.

#### The second flow (unregistered user invited to join group) poses the following problem:

When the user receives an e-mail for joining geonode, the information that the invite is somehow related to joining a group is kept. However, after having registered, since the new account must still be approved by an admin, this information is lost in the extra back and forth via e-email. 

#### Several solutions can be implemented:

1. Drop the *invite unregistered user to join group* feature altogether. This is the least effort approach. IMHO it is also the cleanest. It has the following additional benefits:
  
   - User invitations stay generic enough that they are used only for joining the platform and managed solely with the `django-invitations` app;

   - The flow of joining a group is always the same, instead of requiring an explicit invite acceptance step for new users that does not exist for when already existing users are added to a group

   - There is some code in `geonode.groups` that becomes unnecessary and can be removed from geonode, which results in easier maintenance of the project

   The cost of implementation of this approach is 0.5 days.

2. Bind a function handler to the `user_signed_up` signal that will look up the pending invites stored in `geonode.groups.GroupInvitation`, compare them with the details of the newly registered user and automatically add the user to the correct group(s) if their e-mail address is found on the pending invitations table. This would also entail dropping the additional confirmation step where the user would be asked it she accepts joining the group. This would mean that from a user's point of view both flows of being added to a group become more harmonized, since there is no invitation to join a group to be accepted anymore.

    This approach is straightforward to implement. However it will further the current situation of having two different types of invitations, one managed through `django-invitations` that is used for site-wide invites and another managed by custom geonode logic that is used just for inviting unregistered users to a group.

   The cost of implementation should be 1 day.

3. Similar to option 2. but still provide an explicit confirmation step where the user gets to choose whether or not to accept the group invitation. Implementation will be a bit more involved since this means providing dynamic redirection when user registers. `django-allauth` seems to provide a solution for this. IMHO this approach is less attractive since it continues to treat the act of joining a group in different ways, whereby some users have to accept an invite while others do not.

   This approach should take 1.5 days to implement.

4. Design and implement a new feature in geonode whereby it would become possible to send and access notifications in a centralized registry. These could be made to trigger corresponding actions and be extensible. This could be then used for storing a procedure that would automatically add the user should be added to the group once the corresponding e-mail address were to register. This facility could be useful for a number of additional workflows, such as pushing notifications for long running operations.

    Such a feature could build upon the already used django-activity-stream or pinax-notifications packages, but it would also require some significant design and development efforts. Assessing the cost of implementation for this approach would need a more detailed planning stage, but it would likely take 5+ days

Personally I would go for **option 1. Removing the feature to invite unregistered users to join a group altogether**.

@simboss @afabiani what is your opinion?

**Comment by simboss** [comment]
@ricardogsilva I would go for **1** right away and then discuss with unesco if they want to use some of the support time for **4**.

This gives us a clean path ahead with a short term clean fix and longer term flexible solution. Make also sure that 1 ends up in geonode as a fix.

Before moving forward let's wait for @meyerch feedback.

**Comment by meyerch** [comment]
Hello, 
Thank you for your feedback on this.
We agree with the proposal of going forward with solution #1 to quickly solve the issue, and then think about solution #4 as a future development. 

As for the generic invitation to join the platform, is it possible to display the following text?
![email-invitation.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46ed999b2a8ecea081ffb657a549f99b27046b313916ac217586f76ae5002e580e0b4aa9450e0915f9fb8c7da6b63f708e5c0ae8d76f4d68d47f40f2ecdd42e64bef62994597b4bd804dc5e93e22ef8f8a48371c021d9bb2.PNG)

**Comment by meyerch** [comment]
Exemple of automatic invitation email to be sent. 

[Email subject]: Giuseppe Arduino is inviting you to join her/his group on IHP-WINS

Dear Sir/Madam,

Giuseppe Arduino is inviting you to join the “Theme 5: Ecohydrology” group on http://ihp-wins-dev.geo-solutions.it/ 

To do so, please register at http://ihp-wins-dev.geo-solutions.it/account/signup/. Then, send a message through http://ihp-wins-dev.geo-solutions.it/messages/create/1024/ to Giuseppe to confirm your willingness to join her/his group. 

We look forward to seeing you on the platform,

The IHP-WINS team

[electronic signature as in the picture above]

----------

Madame, Monsieur, 

Giuseppe Arduino vous invite à rejoindre le groupe “Theme 5: Ecohydrology” sur http://ihp-wins-dev.geo-solutions.it/ 

Pour cela, merci de vous enregistrer sur http://ihp-wins-dev.geo-solutions.it/account/signup/. Ensuite, vous pourrez envoyer un message à Giuseppe via http://ihp-wins-dev.geo-solutions.it/messages/create/1024/ pour confirmer votre volonté de rejoindre son groupe.  

Au plaisir de vous croiser sur la plateforme,

L’équipe IHP-WINS

[electronic signature as in the picture above]

**Comment by ricardogsilva** [comment]
Dear @meyerch 

As agreed I've implemented option 1 - remove the group specific invites.
I've also made the necessary changes so that the invitation e-mails follow your requested templates.
I've had to make a slight change in the text though. Instead of

> Giuseppe Arduino is inviting you to join the “Theme 5: Ecohydrology” group on http://ihp-wins-dev.geo-solutions.it/

Since we removed group specific invites, the text is simply:

> Giuseppe Arduino is inviting you to join http://ihp-wins-dev.geo-solutions.it/

In addition, the e-mail also includes a paragraph where the various groups that the inviter is a manager are displayed:

> Giuseppe Arduino manages the following groups:
>  - test-group

Once the changes are approved by @afabiani, please test them out and let me know if the implemented solution is a suitable compromise.

As previously stated we can devise a more involved technical solution in order to bring back group invitations for non registered users in the future, if there is interest from your part.

**Comment by ricardogsilva** [comment]
@afabiani please review this PR together with the changes in branch

https://github.com/geosolutions-it/geonode/tree/remove-non-registered-users-group-invitations

That branch is based on the pending work to add oauth client support into geonode, which hasn't been merged into geonode official yet.

**Comment by meyerch** [comment]
Hi, 
I tried testing the solution but face several problems: 
1/ In the "Invite user" section, I can only enter 1 email address at a time. Whenever I entered 2 or more (whether separated by a semicolon or each on a different line), I get an error message: "Enter a valid email address".
![invite_user.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46e19a8c2a9686969afaba51e611b2843b492d762a10e17132d0ba29bb162e5509091fa9435d0944fcfa892fa6b73976884e17ff933c0120996048e6edc00db513bb27814e80b7ba8d44c4e03824ee8b875d73.PNG)
It is crucial that the user can enter several email addresses at once to invite users to join the platform. Can you implement a solution so that up to 100 email addresses can be entered/100 invitations can be sent at once? With email addresses separated by semicolons, just like when writing a classic email. 

2/ I tried inviting my personal email addresses (meyer_chloe@yahoo.com; meyer_chloe@ymail.com) but never received any invitation on those mailboxes.

**Comment by meyerch** [comment]
To complement point 2/ in the previous message, please find below the text for the pre-written message.
Out of clarification, when a manager "invite users" to join the platform, the future users receives an email. Within this email, we are inviting the future users to register on the platform. We are also inviting them to, once registerd, send an email to the person who invited them  to join the group he/she manages. To do so, the email includes a URL that takes the future user to a 'message page' where he can write to the manager. On this page, the manager is already selected as a recipient of the message. 
So, our idea is too ease even more this process by providing the future-user with a pre-written message when clicking on the URL: unless he wishes to change anything, the future user only has to click "send". 

Here is the text (we made it simple enough so that the exact same thing can be used for all users whatsoever: no mention of particular group/no name/etc.): 
[Subject:] "Request to join your group"
[Text:]"Dear group manager, 

I am now a member of IHP-WINS, and I would like to join the group that you are managing.
Please add me to your group.

Thank you." 

Here is the expected result when the person clicks in the URL contained in the invitation she/he receives. 
![message_to_joinGroup.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46e59189308384acb0fdb07cae52f99a0e5428663946ef696097ec22ba416f43554d4ba5100c5b41f8afdf7aa3b439268a5c08edd56d5e71c03f03b0aac04ee155e57dd205d0f4ae805ec1ef3724e28a884d3c18051f81fb5a32b2d4.PNG)

Let me know if you have any question

**Comment by afabiani** [comment]
@meyerch after struggling several hours on this topic, I should have found a suitable solution for fix/improve this issue in order to match all your 3 requirements.

About the priorities, I'm going to proceed as follows:

1. Multiple email addresses for non-registered users; this is already in progress. Basically I had to override several parts of the "django-invitations" plugin and figure out how to change the Adapters in order to accept more than one email address.

2. Check why emails are not sent; I'm expecting here just a configuration issue.

3. Default subject and message for Group Invitations; this is also should be easy to achieve. I just would like to allow you to be able to customize somehow the text through the Administration panel.

Will send updates on the different points progress.

**Comment by meyerch** [comment]
Hi, 
So tasks #113 and #114 are accepted, thanks. 

As for task #112, (sending invitation to several users at once) it works. However, I have added 2 hopefully easy changes. 
....
....
I have one last change to require, about the invitation email that people will receive:

**Comment by meyerch** [comment]
I have one last general change to require, about the invitation email that people will receive. 
After testing it, I realized that non-manager members of WINS can also invite people to join WINS (which is absolutely normal and ok). So the message needs to be slightly updated in the next (see below). Sorry about that, I should have thought ofit before.

1/ I would like to change the part that was reading "xxx.xxx manages the following group(s):" for "xxx.xxx is a member of the following group(s)". 
This means that in the case of managers, the list include all the groups they are part of whether as a manager or as a simple member. 
As for non-manager users, it means that there is a list of all the group(s) they are a member of. 
For example, for gabin.archambault, the list would say "Theme 6: Water education" and "Theme 2: Groundwater". For najet.guefradj, the list would say"Theme 6: Water education" 
Do you think it is feasible ? 

2/ 

3/  I noticed that some links are misplaced and there are some errors in the text. Below is the definitive version for the text. In brackets are special text, i.e. links or image; I have defined them at the end out of clarification (in the current invitation, only 1 link is incorrect) 
Please display the same thing as what is within the quotation marks (see in the next message a picture of the final result): 

"
English / Français 

Dear Sir/Madam,

gabin.archambault is inviting you to join the Water Information Network System ([ihp-wins-dev.geo-solutions.it]).

To do so, please register at [IHP-WINS Registration]. 

Once you receive the confirmation that your account is activated, you can notify gabin.archambault that you wish to join her/his group(s) through [this link].

gabin.archambault is a member of the following group(s):
•	[IHP-Theme6-Water-education]
•	[IHP-Theme2-Groundwater]

We look forward to seeing you on the platform,

The IHP-WINS team
 [Electronic signature]
________________________________________
Madame, Monsieur,

gabin.archambault vous invite à rejoindre le Système de Réseau d’Information sur l’Eau ([ihp-wins-dev.geo-solutions.it]).

Pour cela, merci de vous enregistrer sur [Inscriptions IHP-WINS]. 

Une fois reçue la confirmation d’activation de votre compte, vous pouvez informer gabin.archambault que vous souhaitez rejoindre son(ses) groupe(s) via [ce lien].  

gabin.archambault est membre du(des) groupe(s) suivant(s):
•	[IHP-Theme6-Water-education]
•	[IHP-Theme2-Groundwater]

Au plaisir de vous croiser sur la plateforme,

L’équipe IHP-WINS
 [electronic signature]"

Here are the links 
[ihp-wins-dev.geo-solutions.it]: ihp-wins-dev.geo-solutions.it (this one is correctly working already in the current invitation)
[IHP-WINS Registration]/[Inscriptions IHP-WINS]: http://ihp-wins-dev.geo-solutions.it/account/signup/  (this link in wrong in the current invitation)
 [this link]/[ce lien]: link to the pre-filled message on WINS with the inviting person as reicpient (this link is correctly working in the current invitation)
[IHP-Theme6-Water-education] & [IHP-Theme2-Groundwater] & etc : link to the designated group. This is working fine in the current invitation.

**Comment by meyerch** [comment]
![invitation_firstHalf.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46e19a8c2a9682bd86e6b17ca254e2873d6e267f2f46ef696097ec22ba416f43554d4ba5100c5b41f8afdf7aa3b439268a5c08edd56d5e71c03f03b0aac04ee155e57dd205d0f4ae805ec2e83724e28a884b3718001f8ff9583db2d4.PNG)
![invitation_secondHalf.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46e19a8c2a9682bd86e6b17cb758f39b27420f722502e1673295f127b54778154d555ca948090b40f9fddf7df7b33e75dc5c0beed1695f75c33b13bea4905dee5fee7fd616cfe3fe9846d2e93a2fe68e8e483a17031d81fa5f33a68b06.PNG)

**Comment by meyerch** [comment]
Accepted
